### PR TITLE
Document boundary conditions

### PR DESCRIPTION
**Pull request by guyer**
Add documentation for Robin and internal boundary conditions, [which](https://www.mail-archive.com/fipy@nist.gov/msg03671.html) [keeps](https://stackoverflow.com/questions/42769749/general-boundary-conditions) [coming](https://www.mail-archive.com/fipy@nist.gov/msg04091.html) [up](https://stackoverflow.com/questions/47169385/how-to-implement-internal-neumann-and-robin-conditions7) (see usnistgov/fipy#426, too)

### CODE REVIEW

File: documentation/USAGE.rst [hunk 1]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
-Applying internal boundary conditions
-=====================================
+Applying Robin boundary conditions
+==================================
+
+The Robin condition
+
+.. math::
+
+   a\phi + b\hat{n}\cdot\nabla\phi = g\qquad\text{on $f=f_0$}
```
Comment (wd15) [inline documentation/USAGE.rst]: Probably use \vec{a} there.

File: documentation/USAGE.rst [hunk 2]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
+
+Internal Robin condition
+------------------------
+
+Nothing different needs to be done when 
+`applying Robin boundary conditions`_ at internal faces.
+
+.. note::
+
+   While we believe these derivations are "correct", they often do not
```
Comment (wd15) [inline documentation/USAGE.rst]: Which derivations and what doesn't work? I'm not sure what you're referring to. Do you mean that "While the derivations above work well for external boundary conditions they fail to work for internal boundary conditions"?

File: documentation/USAGE.rst [hunk 3]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
+   seem to produce the intuitive result. At this point, we think this has 
+   to do with the pathology of "internal" boundary conditions, but remain 
+   open to other explanations. :term:`FiPy` was designed with diffuse 
+   interface treatments (phase field and level set) in mind and, as such, 
+   internal "boundaries" do not come up in our own work and have not 
+   received much attention.
+
+.. note::
+
+  One must be careful to distinguish between constraining internal cell
```
Comment (wd15) [inline documentation/USAGE.rst]: Maybe start off with what the note is actually about and/or make it a warning.

Warning: the constraints mechanism are not designed to constrain internal values for variables that are being solved by equations. In particular, one must be careful

File: documentation/USAGE.rst [hunk 4]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
->>> eqn.solve(var)
->>> print var
-[ 0.25  0.75]
+  >>> eqn.solve(var)
+  >>> print var
+  [ 0.25  0.75]
 
-However, if a constraint is used without the source term constraint an
-unexpected value is obtained
+  However, if a constraint is used without the source term constraint an
```
Comment (wd15) [inline documentation/USAGE.rst]: This is a little confusing, maybe layout two code blocks with all the code for both versions, but with no text interspersed

**Comment by wd15** [comment]
Tests work and the documents build, it looks nice. Just a few changes.

File: documentation/USAGE.rst (hunk 1 above)
Comment (guyer) [reply documentation/USAGE.rst]: Per [wikipedia](https://en.wikipedia.org/wiki/Robin_boundary_condition), it's a scalar equation. For completeness, it should probably be $\hat{n}\cdot\vec{a}$

File: documentation/USAGE.rst (hunk 2 above)
Comment (guyer) [reply documentation/USAGE.rst]: The derivation for how to modify a DiffusionTerm and how to modify a ConvectionTerm. Oh, I see. I didn't mean that to be under "Internal Robin condition". It's a generic issue: I couldn't make that substitute-and-extrapolate approach work for `examples/convection/robin.py`, but I think the generic idea is on the right track for a Robin condition that isn't "just the flux".

File: documentation/USAGE.rst (hunk 3 above)
Comment (guyer) [reply documentation/USAGE.rst]: agreed

File: documentation/USAGE.rst (hunk 4 above)
Comment (guyer) [reply documentation/USAGE.rst]: Not sure I agree. I'm a strong believer in literate programming, but I'll look at making sure the presentation is as clear as possible.

File: documentation/USAGE.rst [hunk 5]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
-``value``. It does not have any effect on the implicit value of ``var`` at the
-``mask`` location during the linear solve so it is not a substitute
-for the source term machinations described above. Future releases of
-:term:`FiPy` may implicitly deal with this discrepancy, but the current
-release does not. A simple example can be used to demonstrate this::
+  simply fixes the returned value of ``var`` at ``mask`` to be
+  ``value``. It does not have any effect on the implicit value of ``var`` at the
+  ``mask`` location during the linear solve so it is not a substitute
+  for the source term machinations described above. Future releases of
+  :term:`FiPy` may implicitly deal with this discrepancy, but the current
```
Comment (tkphd) [inline documentation/USAGE.rst]: If there is not an issue for fixing this, please open one, or remove the time-dependence of the explanation.

File: documentation/USAGE.rst (hunk 1 above)
Comment (guyer) [reply documentation/USAGE.rst]: Converted $a$ to $\hat{n}\cdot\vec{a}$ throughout

File: documentation/USAGE.rst (hunk 2 above)
Comment (guyer) [reply documentation/USAGE.rst]: Clarified that the earlier Robin derivation is believed "correct"

File: documentation/USAGE.rst [hunk 6]
```diff
@@ -535,68 +535,223 @@ Further demonstrations of spatially varying boundary condition can be found
...
+
+The Robin condition can be used to substitute for the expression 
+:math:`\left(\hat{n}\cdot\nabla\phi\right)_{f_0}`
+but we note that :term:`FiPy` calculates variable values at cell centers 
+and gradients at intervening faces. We obtain a first-order approximation 
+for :math:`\left(\hat{n}\cdot\nabla\phi\right)_{f_0}` in terms of 
+neighboring cell values by substituting 
+:math:`\phi_{f_0}
+\approx \phi_P - \left(\vec{d}_{fP}\cdot\nabla\phi\right)_{f_0}
+\approx \phi_P - \left(\hat{n}\cdot\nabla\phi\right)_{f_0}\left(\vec{d}_{fP}\cdot\hat{n}\right)_{f_0}`
```
[the remark this answers is not in the document]
Comment (guyer) [reply documentation/USAGE.rst]: all equations should be inline

File: documentation/USAGE.rst (hunk 5 above)
Comment (guyer) [reply documentation/USAGE.rst]: Filed #581

File: documentation/USAGE.rst (hunk 4 above)
Comment (guyer) [reply documentation/USAGE.rst]: @wd15 I've added some clarify text in f5273cc. Please check.

**Pull request merged**